lib/puppet/provider/cisco_ace/nxapi.rb

```diff
@@ -0,0 +1,222 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'cisco_node_utils' if Puppet.features.cisco_node_utils?
+begin
+  require 'puppet_x/cisco/autogen'
+rescue LoadError # seen on master, not on agent
+  # See longstanding Puppet issues #4248, #7316, #14073, #14149, etc. Ugh.
+  require File.expand_path(File.join(File.dirname(__FILE__), '..', '..', '..',
+                                     'puppet_x', 'cisco', 'autogen.rb'))
+end
+
+Puppet::Type.type(:cisco_ace).provide(:nxapi) do
+  desc 'The NXAPI provider.'
+
+  confine feature: :cisco_node_utils
+  defaultfor operatingsystem: :nexus
+
+  mk_resource_methods
+
+  # Property symbol array for method auto-generation.
+  #
+  # NOTE: For maintainability please keep this list in alphabetical order and
+  # one property per line.
+  ACL_NON_BOOL_PROPS = [
+    :action,
+    :proto,
+    :src_addr,
+    :src_port,
+    :dst_addr,
+    :dst_port,
+    :option_format,
+  ]
+
+  ACL_BOOL_PROPS = [
+  ]
+
+  ACL_ALL_PROPS = ACL_NON_BOOL_PROPS + ACL_BOOL_PROPS
+
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:non_bool, self, '@ace',
+                                            ACL_NON_BOOL_PROPS)
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:bool, self, '@ace',
+                                            ACL_BOOL_PROPS)
+
+  def initialize(value={})
+    super(value)
+    afi = @property_hash[:afi]
+    acl_name = @property_hash[:acl_name]
+    seqno    = @property_hash[:seqno]
+    @ace = Cisco::Ace.aces[afi][acl_name][seqno] unless acl_name.nil? && seqno.nil?
+    @property_flush = {}
+  end
+
+  def self.properties_get(afi, acl_name, seqno, instance)
+    debug "Checking acl instance, #{afi} #{acl_name} #{seqno}"
+    current_state = {
+      name:     "#{afi} #{acl_name} #{seqno}",
+      acl_name: acl_name,
+      seqno:    seqno,
+      ensure:   :present,
+      afi:      afi,
+    }
+
+    # Call node_utils getter for each property
+    ACL_NON_BOOL_PROPS.each do |prop|
+      current_state[prop] = instance.send(prop)
+    end
+    ACL_BOOL_PROPS.each do |prop|
+      val = instance.send(prop)
+      if val.nil?
+        current_state[prop] = nil
+      else
+        current_state[prop] = val ? :true : :false
+      end
+    end
+    debug current_state
+    new(current_state)
+  end # self.properties_get
+
+  def self.instances
+    ace_hash = []
+    Cisco::Ace.aces.each do |afi, acl_aces|
+      acl_aces.each do |acl_name, aces|
+        aces.each do |seqno, ace_instance|
+          ace_hash << properties_get(afi, acl_name, seqno, ace_instance)
+        end
+      end
+    end
+    ace_hash
+  end
+
+  def self.prefetch(resources)
+    ace_instances = instances
+    resources.keys.each do |name|
+      provider = ace_instances.find do |ace|
+        resources[name][:seqno] == ace.seqno.to_i &&
+        resources[name][:acl_name] == ace.acl_name.to_s &&
+        resources[name][:afi] == ace.afi.to_s
```

Comment:
Nit:  I would order these checks by the way they are ordered in the hash, ie: check afi, then acl_name, then seqno.